Adding an import in a later cell nullifies variables defined in earlier cells

## Summary

*Generated by Claude*

When a variable is defined in one cell and then new `import` statements are evaluated in a subsequent cell, the variable's value is reset to `null`. Any code in later cells that references that variable then fails with a `NullPointerException`. Combining all imports into a single upfront cell avoids the problem, but that is an unreasonable constraint for interactive notebook use.

## Steps to Reproduce

Run the following cells in order:

**Cell 1** — initial import
```java
import io.bootique.*;
```

**Cell 2** — define a variable
```java
var bq = Bootique.app("-c", "my.yml")
    .autoLoadModules()
    .createRuntime();
```

**Cell 3** — add more imports in a new cell
```java
import io.bootique.config.jackson.*;
import io.bootique.config.*;
```

**Cell 4** — use the variable defined in Cell 2
```java
JsonConfigurationFactory cf =
    (JsonConfigurationFactory) bq.getInstance(ConfigurationFactory.class);
```

## Actual Behavior

Cell 4 throws a `NullPointerException` because `bq` is `null`:

```
java.lang.RuntimeException: java.lang.NullPointerException, Cannot invoke
"io.bootique.BQRuntime.getInstance(java.lang.Class)" because "REPL.$JShell$30B.bq" is null
|       at org.dflib.jjava.kernel.execution.CodeEvaluator.evalSingle(CodeEvaluator.java:145)
|       ...
|   Caused by: jdk.jshell.EvalException: Cannot invoke
"io.bootique.BQRuntime.getInstance(java.lang.Class)" because "REPL.$JShell$30B.bq" is null
|       at .(<Anonymous>#35:1)
```

The internal REPL class name (`$JShell$30B.bq`) reveals that JShell has re-created the snippet class for `bq`'s enclosing context, leaving the field uninitialized.

## Expected Behavior

Adding imports after a variable has been defined should not affect the value of that variable. The notebook's interactive contract is that each cell runs in the existing state; importing new classes is purely additive and should not invalidate prior bindings.

## Workaround

Placing all `import` statements in a single cell *before* defining any variables avoids the issue:

```java
// Cell 1 — all imports together
import io.bootique.*;
import io.bootique.config.jackson.*;
import io.bootique.config.*;
```

```java
// Cell 2 — variable definition (safe now)
var bq = Bootique.app("-c", "my.yml")
    .autoLoadModules()
    .createRuntime();
```

This works, but is impractical for real notebook workflows where imports are naturally discovered and added incrementally.

--- a/jjava-distro/src/test/java/org/dflib/jjava/distro/KernelExecutionIT.java
+++ b/jjava-distro/src/test/java/org/dflib/jjava/distro/KernelExecutionIT.java
@@ -1,0 +1,25 @@
+package org.dflib.jjava.distro;
+
+import org.junit.jupiter.api.Test;
+import org.testcontainers.containers.Container;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class KernelExecutionIT extends ContainerizedKernelCase {
+
+    /**
+     * @see <a href="https://github.com/dflib/jjava/issues/119">#119</a>
+     */
+    @Test
+    public void variableSurvivesLaterImports() throws Exception {
+        String snippet = "%load " + CONTAINER_RESOURCES + "/nullifying_import.jshell";
+
+        Container.ExecResult snippetResult = executeInKernel(snippet);
+
+        assertEquals(0, snippetResult.getExitCode(), snippetResult.getStdout());
+        assertThat(snippetResult.getStdout(), not(containsString("|")));
+    }
+}

--- a/jjava-distro/src/test/resources/nullifying_import.jshell
+++ b/jjava-distro/src/test/resources/nullifying_import.jshell
@@ -1,5 +1,5 @@
-%maven tools.jackson.core:jackson-databind:3.1.2
-import tools.jackson.databind.*; import tools.jackson.databind.json.*;
-var om = JsonMapper.builder().findAndAddModules().build();
-import tools.jackson.databind.node.*; import tools.jackson.databind.type.*;
+%maven com.fasterxml.jackson.core:jackson-databind:2.21.2
+import com.fasterxml.jackson.databind.*;
+var om = new ObjectMapper().findAndRegisterModules();
+import com.fasterxml.jackson.databind.node.*; import com.fasterxml.jackson.databind.type.*;
 om.getClass().getName()

--- a/jjava-distro/src/test/resources/nullifying_import.jshell
+++ b/jjava-distro/src/test/resources/nullifying_import.jshell
@@ -1,0 +1,5 @@
+%maven tools.jackson.core:jackson-databind:3.1.2
+import tools.jackson.databind.*; import tools.jackson.databind.json.*;
+var om = JsonMapper.builder().findAndAddModules().build();
+import tools.jackson.databind.node.*; import tools.jackson.databind.type.*;
+om.getClass().getName()

--- a/jjava-kernel/src/main/java/org/dflib/jjava/kernel/execution/JJavaExecutionControl.java
+++ b/jjava-kernel/src/main/java/org/dflib/jjava/kernel/execution/JJavaExecutionControl.java
@@ -76,6 +76,11 @@ class JJavaExecutionControl extends DirectExecutionControl {
         executor.shutdownNow();
     }
 
+    @Override
+    public void redefine(ClassBytecodes[] cbcs) {
+        loaderDelegate.classesRedefined(cbcs);
+    }
+
     public Object takeResult(String id) {
         Object result = this.results.remove(id);
         if (result == null) {

